content/learning-paths/smartphones-and-mobile/function-multiversioning/semantics.md

```diff
@@ -0,0 +1,23 @@
+---
+title: Semantics
+weight: 2
+
+### FIXED, DO NOT MODIFY
+layout: learningpathall
+---
+
+## Semantics
+Function Multiversioning allows compiling several implementations of a function into the same binary and then selecting the most appropriate version at runtime. The intention is to take advantage of hardware features for accelerating your application at function level granularity.
+
+To specify a function version you can annotate its declaration with either of  `__attribute__((target_version("name")))` or `__attribute__((target_clones("name",...)))`, where "name" denotes one or more architectural features separated by '+'. This annotation implies a dependency between a function version and the feature set it is specified for. The compiler generates optimized versions of the function for the specified targets. The `target_clones` attribute behaves just like `target_version` but specifies multiple versions for the same function definition. The former is perhaps suitable for functions that the compiler can optimize differently depending on the requested features, whereas the latter allows the user to manually optimize a version at the source level using intrinsics or inline assembly.
+
+A hardware platform may support multiple architectural features from the dependency sets, or it may not support any. Therefore Function Multiversioning provides a convenient way to select the most appropriate version of a function at runtime. The selection is permanent for the lifetime of the process and works as follows:
+- Select the most specific version (the one with most features), else
```

Comment:
This description is imprecise, and we're planning to clarify and slightly modify the prioritization criteria anyway.